amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/fanout/FanOutRecordsPublisher.java

```diff
@@ -158,20 +151,26 @@ RecordFlow evictAckedEventAndScheduleNextEvent(RecordsDeliveryAck recordsDeliver
         // RecordFlow of the current event that needs to be returned
         RecordFlow flowToBeReturned = null;
 
+        final RecordsRetrieved recordsRetrieved = recordsRetrievedContext != null ?
+                recordsRetrievedContext.getRecordsOrShutDownEvent()
+                        .map(recordsEvent -> recordsEvent, shutDownEvent -> null) : null;
+
         // Check if the ack corresponds to the head of the delivery queue.
-        if (recordsRetrievedContext != null && recordsRetrievedContext.getRecordsRetrieved().batchUniqueIdentifier()
+        if (recordsRetrieved != null && recordsRetrieved.batchUniqueIdentifier()
                 .equals(recordsDeliveryAck.batchUniqueIdentifier())) {
             // It is now safe to remove the element
             recordsDeliveryQueue.poll();
             // Take action based on the time spent by the event in queue.
             takeDelayedDeliveryActionIfRequired(shardId, recordsRetrievedContext.getEnqueueTimestamp(), log);
             // Update current sequence number for the successfully delivered event.
-            currentSequenceNumber = ((FanoutRecordsRetrieved)recordsRetrievedContext.getRecordsRetrieved()).continuationSequenceNumber();
+            currentSequenceNumber = ((FanoutRecordsRetrieved)recordsRetrieved).continuationSequenceNumber();
             // Update the triggering flow for post scheduling upstream request.
             flowToBeReturned = recordsRetrievedContext.getRecordFlow();
-            // Try scheduling the next event in the queue, if available.
+            // Try scheduling the next event in the queue or execute the subscription shutdown action.
             if (!recordsDeliveryQueue.isEmpty()) {
-                scheduleNextEvent(recordsDeliveryQueue.peek().getRecordsRetrieved());
+                recordsDeliveryQueue.peek().getRecordsOrShutDownEvent()
+                        .apply(recordsEvent -> scheduleNextEvent(recordsEvent),
```

Comment:
Can this also move out into the RecordsRetrievedContext? each event would define execution action, similar to the shutdownEvent's getSubscriptionShutdownAction, and this flow would just call executeEvenAction independent of the event type.